### ISSUE TITLE
OneDrive sync: Can't transfer non file/directory

### ISSUE THREAD

**Issue by andrewhollomon** [ncw/rclone#614]
Hi, when trying to sync a folder to OneDrive, running into 'Can't transfer non file/directory', only on some of the files and folders in the source folder. Other files and folders successfully sync. I can't see an obvious pattern relating to the files that do not sync.

-Rclone 1.32
-Windows Server 2012 r2 (64bit)
-OneDrive Personal

2016/08/12 10:10:08 rclone: Version "v1.32" starting with parameters ["C:\\ProgramData\\RClone\\rclone.exe" "--config" "c:\\programdata\\rclone\\.rclone.conf" "sync" "\\\\abc\\data\\ABCtechnology\\Security" "abc-OneDrive:\\Disaster Recovery\\Security" "-v" "--exclude-from" "c:\\programdata\\rclone\\ZGF_FileExclusions.txt" "-q" "--log-file=C:\\ProgramData\\Rclone\\ABC_RcloneErrorLog.txt"]
...
2016/08/12 10:10:09 ABC-OneDrive: Saving new token in config file
2016/08/12 10:10:10 One drive root 'Disaster Recovery/Security': Modify window is 1s
2016/08/12 10:10:10 One drive root 'Disaster Recovery/Security': Reading ""
2016/08/12 10:10:10 KeePass-2.34-Portable.zip: Can't transfer non file/directory
2016/08/12 10:10:10 KeePass-2.34-Setup.zip: Can't transfer non file/directory
...
2016/08/12 10:10:13 Go routines at exit 2

**Comment by ncw** [comment]
Are they all .zip files maybe?  I wonder if you have some extension to Windows which is treating these in a special way which confuses rclone.

Can you run this program with the full path to one of those files and send me the output please?

http://pub.rclone.org/stat-windows-386.zip

Eg

```
C> stat.exe "KeePass-2.34-Portable.zip"
```

Thanks

**Comment by andrewhollomon** [comment]
Sure, here is the output from stat.exe, for a failing file:

Name = "KeePass-2.34-Portable.zip"
Size = 2923638
Mode = 0x80001B6
ModTime = 2016-07-26 10:57:08.1369745 -0700 PDT
IsDir = false
Sys = &{1568 {448576140 30533479} {448576140 30533479} {543266961 30533479} 0 2923638}

and for a file that does sync properly:

Name = "ABC.txt"
Size = 1165
Mode = 0x1B6
ModTime = 2013-03-04 11:19:02.0991664 -0800 PST
IsDir = false
Sys = &{32 {687528094 29898661} {1807448354 30390944} {555399344 30284045} 0 1165}

The files that generate Can't transfer non file/directory are a mix of zip and dll files.

**Comment by ncw** [comment]
Hmm, those mode flags indicate that that zip file is a symlink

Looking at the go source I see

```go
	if fs.sys.FileAttributes&syscall.FILE_ATTRIBUTE_REPARSE_POINT != 0 {
		m |= ModeSymlink
	}
```

And according to the msdn docs

FILE_ATTRIBUTE_REPARSE_POINT 1024 (0x400): A file or directory that has an associated [reparse point](https://msdn.microsoft.com/en-us/library/windows/desktop/aa365503(v=vs.85).aspx), or a file that is a symbolic link.

A bit more digging [reveals this](http://answers.microsoft.com/en-us/windows/forum/windows_7-performance/what-is-a-reparse-point-can-anyone-reveal-the/17b9b457-6c8a-4e83-a445-e603011a6b95) - the second answer says

> I also use 64-bit Windows 7 and generally love it, but was stumped for awhile with this same backup problem. Apparently reparse points are related to or associated with backup restore points in some manner. I had moved files to a new physical disk and partition I installed recently. Backups of the moved files were failing with the 0x81000037 error. I followed the steps below and was able to get a successful backup.

  1. Open Windows Explorer and right click the drive letter where the files are stored that cannot be backed up. Click Properties.
  2. From the General tab click "Disk Cleanup".
  3. On the Disk Cleanup window, switch to the More Options tab.
  4. In the System Restore and Shadow Copies section, click "Clean up..."
  5. Click Delete

> This initially made me nervous about possible file deletion, so before testing I made another copy of the files to a different partition - a backup of my backup I guess. However this process apparently only deletes restore point information associated with files on a given partition. Once that was done, backups worked fine.

So you could try that and see if it improves things.

I could also ignore the ModeSymlink on windows which may or may not do the right thing, I'm not sure!

**Comment by andrewhollomon** [comment]
Thanks, based on what you discovered about reparse points; some additional information that may be relevant is that the zip file in question lives on a volume with [Microsoft Server's native volume deduplication](https://msdn.microsoft.com/en-us/library/hh769303(v=vs.85).aspx) feature enabled.

The same volume is also protected using Microsoft Shadow Copies. I am not entirely sure how reparse points figure in with shadow copies. But, unlike a locally protected volume, in our case the shadow copy data is stored on another volume entirely and managed by a Microsoft Data Protection Manager (DPM) server.

Unfortunately for us, it is not an option to disable duplication or shadow copies on the volume in question, but I could certainly try using rclone to sync from a more normal volume (no shadow copy protection, no native dedupe) to verify that works as expected.

**Comment by andrewhollomon** [comment]
Synching the same data from another volume without Microsoft's volume deduplication enabled results in no errors.

**Comment by ncw** [comment]
Interesting! When I get back from vacation I'll send you a build which ignores the junction point and we can see what happens.

**Comment by ncw** [comment]
Here is a beta with a potential fix: http://pub.rclone.org/v1.33-8-ga02273e%CE%B2/

Let me know how you get on!

Thanks

**Comment by ncw** [comment]
Did this work?  (Note to self - this is in the `ingore-junction-points` branch!)

**Comment by andrewhollomon** [comment]
Hi Nick, yes it did work - all files synched from the de-duped volume using
the build you sent me. I think I posted to the original thread. Thanks,
Andrew

**Comment by ncw** [comment]
Thanks for checking - I'll merge this to master now

**Comment by ncw** [comment]
@andrewhollomon not that I've changed the way this works as the original fix causes different problems (see #2336), so now to backup files with reparse points you'll have to use the `-L` flag.  This change will be in effect from v1.42 which I'm just about to release.

This will have the unfortunate effect that if you are using `rclone sync` with v1.42 and  you don't use `-L` then it will delete the files with reparse points from the backup, so please add the `-L` to your script before trying v1.42.  Also try first with --dry-run!

Thanks
